### ISSUE TITLE
Request to implement SCALE optimizer

### ISSUE THREAD

**Issue by jim-plus** [pytorch/torchtune#2928]
Although the proposed SCALE optimizer was only validated for pretraining, the claimed memory savings make it appealing for a fine-tuning optimizer as well.

https://arxiv.org/abs/2506.16659

**Comment by krammnic** [comment]
Hey, thank you for the issue, would you like to try to implement it by yourself?

**Comment by jim-plus** [comment]
Optimizers aren't my forte, but I could give it a look and at least share what I've found.

This appears to be a reference implementation of SCALE:
https://github.com/OptimAI-Lab/Minimalist_LLM_Pretraining

This appears to be the optimizer code itself:
https://github.com/OptimAI-Lab/Minimalist_LLM_Pretraining/blob/main/mem_eff_pt/pt_scale/scale_optimizer.py

They provide example hyperparameters:

torchrun --standalone --nproc_per_node 4 torchrun_main_DDP.py \
    --model_name 350m_scale \
    --model_config configs/llama_350m.json \
    --optimizer scale \
    --lr 1e-3 \
    --momentum 0.9 \
    --weight_decay 0.0 \
    --batch_size 128 \
    --total_batch_size 512 \
    --num_training_steps 60000 \
    --warmup_steps 6000 \
    --dtype bfloat16 \
    --eval_every 1000 \
    --save_every 99999 \
    --seed 42  \
    --scheduler cosine \
    --dataset_path /path/to/c4/en \

**Comment by jim-plus** [comment]
The optimizer seems mostly straightforward to port except for the problem of how to identify if the layer being operated on is the last layer.

**Comment by krammnic** [comment]
@jim-plus I think it should be quite easy with the nature of the `TransformerDecoder`. If you are still interested let's try to wrap up PR for this together (I will help you with interaction with internals)!

**Comment by jim-plus** [comment]
I've come across a revision to the preprint which they directly did a comparison against Muon and found SCALE to be competitive up to 1B parameters, and for pretraining on a Llama 7B model given sufficient steps. https://arxiv.org/abs/2506.16659v2

I think it would be interesting to see how well SCALE performs for PEFT or for continued pretraining on more representative open weight models, though perhaps more evidence is required before any PR is actually merged. What should I ramp up on?

**Comment by krammnic** [comment]
@jim-plus Yep, I agree that we need slightly more evidence. You can try to explore this topic more by yourself (but for the PR only post-training domain is relevant) and if you can show that it is representative enough we will be able to merge the PR which you will open as a consequence.
